[Homework Questions Link](https://github.com/DataTalksClub/stock-markets-analytics-zoomcamp/blob/main/cohorts/2026/homework2.md)

## Question 1: [IPO] Withdrawn IPOs by Company Type

### What is the total withdrawn IPO value (in $ millions) for the company class with the highest total withdrawal value?

In [ ]:
import pandas as pd
import re
import numpy as np

In [ ]:
url = "https://www.iposcoop.com/ipos-recently-filed/"
ipos_recently_filed = pd.read_html(url)[0]

# look at the ipo data
ipos_recently_filed.head(3)

,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating
0,2026-09-28,Encore Medical Inc.,EMI,The Oak Ridge Financial Services Group/Network...,3.0,$5.00,$5.00,$15.00,Wednesday,S/O
1,2026-09-25,Elevation Acquisition Group,ELEVU,Maxim Group,10.0,$10.00,$10.00,$100.00,TBA,S/O
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.0,$15.00,$18.00,$825.00,Withdrawn,S/O


In [ ]:
ipos_recently_filed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   File Date             500 non-null    object 
 1   Company               500 non-null    object 
 2   Symbol                500 non-null    object 
 3   Managers              500 non-null    object 
 4   Shares (millions)     500 non-null    float64
 5   Price Low             464 non-null    object 
 6   Price High            464 non-null    object 
 7   Est $ Vol (millions)  495 non-null    object 
 8   Expected To Trade     500 non-null    object 
 9   SCOOP Rating          500 non-null    object 
dtypes: float64(1), object(9)
memory usage: 39.2+ KB


In [ ]:
# filter to 'Withdrawn' only for Expected to Trade
withdrawn_df = ipos_recently_filed[ipos_recently_filed['Expected To Trade']=='Withdrawn']
display(withdrawn_df.shape)

(35, 10)

In [ ]:
# classifies company type based on keywords
def classify_company_type(company_name):
    company_name_lower = str(company_name).lower()
    if 'technologies' in company_name_lower:
        return 'Technologies'
    elif 'acquisition corp' in company_name_lower or \
         'acquisition corporation' in company_name_lower or \
         re.search(r'\bcorp\b', company_name_lower):
        return 'Acquisition Corp'
    elif 'inc' in company_name_lower or 'incorporated' in company_name_lower:
        return 'Inc.'
    elif 'group' in company_name_lower:
        return 'Group'
    elif 'ltd' in company_name_lower or 'limited' in company_name_lower:
        return 'Limited'
    elif 'holdings' in company_name_lower or 'holding' in company_name_lower:
        return 'Holdings'
    else:
        return 'Other'

withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)
withdrawn_df.head(3)

/tmp/ipykernel_1996/1360166711.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Company Type'] = withdrawn_df['Company'].apply(classify_company_type)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group


In [ ]:
def find_avg_price(row: pd.DataFrame):
  low = pd.to_numeric(str(row["Price Low"]).replace('$', ''), errors="coerce")
  high = pd.to_numeric(str(row["Price High"]).replace('$', ''), errors="coerce")

  return (low + high) / 2

withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)
withdrawn_df.head(3)

/tmp/ipykernel_1996/597125063.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Avg_price'] = withdrawn_df.apply(find_avg_price, axis=1)


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,$825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,$7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,$16.88,Withdrawn,S/O,Group,4.5


In [ ]:
withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")
withdrawn_df.head(3)

/tmp/ipykernel_1996/761777133.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df["Shares (millions)"] = pd.to_numeric(withdrawn_df['Shares (millions)'], errors="coerce")
/tmp/ipykernel_1996/761777133.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Est $ Vol (millions)'] = pd.to_numeric(withdrawn_df['Est $ Vol (millions)'].str.replace('$', '').replace(',', ''), errors="coerce")


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5


In [ ]:
withdrawn_df['Shares_offered_value'] = np.where(
    pd.isna(withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']),
    withdrawn_df['Est $ Vol (millions)'],
    withdrawn_df['Shares (millions)'] * withdrawn_df['Avg_price']
)
withdrawn_df.head()

/tmp/ipykernel_1996/804680285.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  withdrawn_df['Shares_offered_value'] = np.where(


,File Date,Company,Symbol,Managers,Shares (millions),Price Low,Price High,Est $ Vol (millions),Expected To Trade,SCOOP Rating,Company Type,Avg_price,Shares_offered_value
2,2026-09-25,Holtec Nuclear (Withdrawn),HNUC,J.P.Morgan/Guggenheim Securities/Goldman Sachs...,50.00,$15.00,$18.00,825.00,Withdrawn,S/O,Other,16.5,825.000
24,2026-09-18,Rise Smart Group (Withdrawn),RSHL,American Trust Investment Services/Prime Numbe...,1.80,$4.00,$4.00,7.20,Withdrawn,S/O,Group,4.0,7.200
27,2026-09-17,C2 Capital Group (Withdrawn),CCLV,Dominari Securities,3.75,$4.00,$5.00,16.88,Withdrawn,S/O,Group,4.5,16.875
39,2026-09-15,OTSAW Ltd. (Withdrawn),OTSA,Aegis Capital,4.40,$4.50,$5.50,22.00,Withdrawn,S/O,Limited,5.0,22.000
45,2026-09-10,Motive Technologies (Withdrawn),MTVE,J.P.Morgan/Citigroup/Barclays/Jefferies/RBC Ca...,0.00,NaN,NaN,100.00,Withdrawn,S/O,Technologies,NaN,100.000


In [ ]:
agg_df = withdrawn_df.groupby(['Company Type']) \
.Shares_offered_value.agg(['sum']) \
.rename({'sum': 'sum_shares_offered_value'}, axis=1)
agg_df

,sum_shares_offered_value
Company Type,
Acquisition Corp,499.9850
Group,56.5750
Holdings,311.6575
Inc.,351.0000
Limited,219.2500
Other,1115.4450
Technologies,184.9000


In [ ]:
agg_df.sort_values(by='sum_shares_offered_value', ascending=False)

,sum_shares_offered_value
Company Type,
Other,1115.4450
Acquisition Corp,499.9850
Inc.,351.0000
Holdings,311.6575
Limited,219.2500
Technologies,184.9000
Group,56.5750


### Answer: The class 'Acquisition Corp' has the highest total value of withdrawals (outside of Other) with $499.9850.

## Question 2: [IPO] Median Sharpe Ratio for 2025 IPOs (First 8 Months)

### What is the median Sharpe ratio (as of 11 September 2026) for companies that went public before 1 September 2025?

In [238]:
import pandas as pd
from time import sleep
import yfinance as yf
import numpy as np
import datetime as dt

In [239]:
url = "https://www.iposcoop.com/2025-pricings/"
ipos25 = pd.read_html(url)[0]
display(ipos25.head(3))
display(ipos25.info())

# convert datatypes to correct format
ipos25["Offer Date"] = pd.to_datetime(ipos25["Offer Date"])
ipos25["Return"] = pd.to_numeric(ipos25["Return"].str.replace("%", ''), errors="coerce")

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating
0,Vine Hill Capital Investment Corp. II,VHCPU,Blank Check,12/18/2025,20.0,$10.00,$0.00,$0.00,0.00%,S/O
1,Andersen Group,ANDG,Consumer Services,12/17/2025,11.0,$16.00,$23.50,$23.79,48.69%,S/O
2,Medline Inc.,MDLN,Health Care,12/17/2025,216.0,$29.00,$41.00,$43.83,51.14%,S/O


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Company            231 non-null    object 
 1   Symbol             231 non-null    object 
 2   Industry           231 non-null    object 
 3   Offer Date         231 non-null    object 
 4   Shares (millions)  231 non-null    float64
 5   Offer Price        231 non-null    object 
 6   1st Day Close      231 non-null    object 
 7   Current Price      231 non-null    object 
 8   Return             231 non-null    object 
 9   SCOOP Rating       231 non-null    object 
dtypes: float64(1), object(9)
memory usage: 18.2+ KB


None

In [240]:
# filter to only before sep 1 2025 offer date and remove those with 0% return
filtered_df = ipos25[(ipos25["Offer Date"] < "2025-09-01") & (ipos25["Return"] != 0)]
# filtered_df = ipos25[ipos25["Offer Date"] < "2025-09-01"]
# filtered_df = filtered_df[filtered_df["Return"] != 0.0]
display(filtered_df.head(3))
display(filtered_df.shape)

,Company,Symbol,Industry,Offer Date,Shares (millions),Offer Price,1st Day Close,Current Price,Return,SCOOP Rating
68,"Picard Medical, Inc.",PMI,Health Care,2025-08-29,4.3,$4.00,$4.59,$1.68,-58.00,S/O
69,GrowHub Limited (The),TGHL,Technology,2025-08-28,3.7,$4.00,$3.38,$0.40,-90.00,S/O
70,TryHard Holdings Limited,THH,Consumer Services,2025-08-28,1.5,$4.00,$6.02,$29.47,636.75,S/O


(146, 10)

In [258]:
# get daily stocks prices
tickers = filtered_df['Symbol'].unique()
stocks_df = pd.DataFrame({'A': []})
start = dt.date(year=2024, month=1, day=1)
end = dt.date(year=2026, month=9, day=15)
risk_free_rate_of_return = 0.05
delisted = []

for i, t in enumerate(tickers):
  # print(i, t)
  try:
    # get data for the current ticker
    temp = yf.Ticker(t).history(start=start, end=end, interval="1d")

    # transformations
    temp['ticker'] = t
    temp['Date'] = pd.to_datetime(temp.index.date)
    temp['year'] = temp.index.year
    temp['month'] = temp.index.month
    temp['weekday'] = temp.index.weekday
    temp['close_252d'] = temp.Close.shift(252)
    # 1 year of trading day's growth
    temp['growth_252d_raw'] = temp.Close / temp['close_252d']
    temp['growth_252d'] = (temp.Close / temp['close_252d']) - 1
    # annualized volatility
    temp['volatility'] = temp.Close.rolling(30).std() * np.sqrt(252)
    # sharpe ratio - risk-adjusted performance of an investment (higher is better)
    temp['sharpe'] = (
        temp['growth_252d'] - risk_free_rate_of_return
        ) / temp['volatility']

  except Exception as e:
    print(f"Skipped {t} due to {e}")
    delisted.append(t)
    continue

  else:
    # save into main dataframe
    if stocks_df.empty:
      stocks_df = temp
    else:
      stocks_df = pd.concat([stocks_df, temp], ignore_index=True)

  sleep(1)

ERROR:yfinance:$MJID: possibly delisted; no timezone found


Skipped MJID due to 'Index' object has no attribute 'date'


ERROR:yfinance:$EMPG: possibly delisted; no timezone found


Skipped EMPG due to 'Index' object has no attribute 'date'


ERROR:yfinance:$CAEP: possibly delisted; no timezone found


Skipped CAEP due to 'Index' object has no attribute 'date'


ERROR:yfinance:$PTNM: possibly delisted; no timezone found


Skipped PTNM due to 'Index' object has no attribute 'date'


ERROR:yfinance:$AHL: possibly delisted; no timezone found
ERROR:yfinance:$CEPT: possibly delisted; no timezone found


Skipped AHL due to 'Index' object has no attribute 'date'
Skipped CEPT due to 'Index' object has no attribute 'date'


ERROR:yfinance:$SDM: possibly delisted; no timezone found


Skipped SDM due to 'Index' object has no attribute 'date'


ERROR:yfinance:$TBH: possibly delisted; no timezone found


Skipped TBH due to 'Index' object has no attribute 'date'


ERROR:yfinance:$WGRX: possibly delisted; no timezone found


Skipped WGRX due to 'Index' object has no attribute 'date'


ERROR:yfinance:$AGH: possibly delisted; no timezone found


Skipped AGH due to 'Index' object has no attribute 'date'


ERROR:yfinance:$EPWK: possibly delisted; no timezone found


Skipped EPWK due to 'Index' object has no attribute 'date'


ERROR:yfinance:$MTSR: possibly delisted; no timezone found


Skipped MTSR due to 'Index' object has no attribute 'date'


ERROR:yfinance:$SKBL: possibly delisted; no timezone found
ERROR:yfinance:$MCTR: possibly delisted; no timezone found


Skipped SKBL due to 'Index' object has no attribute 'date'
Skipped MCTR due to 'Index' object has no attribute 'date'


In [259]:
stocks_filtered = stocks_df[stocks_df.Date == '2026-09-11']
display(stocks_filtered.describe())

,Open,High,Low,Close,Volume,Dividends,Stock Splits,Date,year,month,weekday,close_252d,growth_252d_raw,growth_252d,volatility,sharpe,Capital Gains
count,132.000000,132.000000,132.000000,132.000000,1.320000e+02,132.0,132.0,132,132.0,132.0,132.0,130.000000,130.000000,130.000000,131.000000,130.000000,1.0
mean,12.831901,13.181377,12.510782,12.800431,1.248240e+06,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,67.044147,1.057613,0.057613,27.573914,inf,0.0
min,0.116000,0.129000,0.115000,0.125000,0.000000e+00,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,0.763000,0.001005,-0.998995,0.000000,-2.534214,0.0
25%,2.137500,2.257500,2.075500,2.098750,2.297500e+04,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,5.550000,0.141035,-0.858965,2.332309,-0.171847,0.0
50%,4.505000,4.790000,4.475000,4.565000,1.708500e+05,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,14.565000,0.594534,-0.405466,7.970052,-0.037267,0.0
75%,17.397501,17.471250,16.732501,17.224999,9.976250e+05,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,35.385000,1.038821,0.038821,23.452319,-0.000343,0.0
max,120.500000,120.500000,120.425003,120.425003,2.168250e+07,0.0,0.0,2026-09-11 00:00:00,2026.0,9.0,4.0,3264.000000,33.638270,32.638270,627.698106,inf,0.0
std,19.213619,19.564220,18.868847,19.135136,3.184362e+06,0.0,0.0,NaN,0.0,0.0,0.0,301.607052,3.049007,3.049007,66.859284,NaN,NaN


### Answer: The median Sharpe ratio is -0.037267 or -0.04.